app/src/editor/canvas/components/CodeEditor/index.jsx

```diff
@@ -1,101 +1,114 @@
-import React from 'react'
+import React, { useState, useCallback } from 'react'
 
 import CodeEditorWindow from './CodeEditorWindow'
 import DebugWindow from './DebugWindow'
 
-export default class CodeEditor extends React.Component {
-    state = {
-        editorOpen: false,
-        debugOpen: false,
-        editorPosition: {
-            x: 0,
-            y: 0,
-        },
-        debugPosition: {
-            x: 0,
-            y: 0,
-        },
-    }
+export const CodeEditor = ({
+    children,
+    code,
+    readOnly,
+    onApply,
+    onChange,
+    debugMessages,
+    onClearDebug,
+}) => {
+    const [editorOpen, setEditorOpen] = useState(false)
+    const [debugOpen, setDebugOpen] = useState(false)
+    const [editorLayout, setEditorLayout] = useState({
+        x: 0,
+        y: 0,
+        width: 600,
+        height: 400,
+    })
+    const [debugLayout, setDebugLayout] = useState({
+        x: 0,
+        y: 0,
+        width: 600,
+        height: 400,
+    })
+    const onShowEditor = useCallback(() => {
+        setEditorOpen(true)
+    }, [setEditorOpen])
+    const onCloseEditor = useCallback(() => {
+        setEditorOpen(false)
+    }, [setEditorOpen])
+    const onShowDebug = useCallback(() => {
+        setDebugOpen(true)
+    }, [setDebugOpen])
+    const onCloseDebug = useCallback(() => {
+        setDebugOpen(false)
+    }, [setDebugOpen])
+    const onEditorPositionUpdate = useCallback((x, y) => {
+        setEditorLayout((layout) => ({
+            ...layout,
+            x,
+            y,
+        }))
+    }, [setEditorLayout])
+    const onEditorSizeUpdate = useCallback((width, height) => {
+        setEditorLayout((layout) => ({
+            ...layout,
+            width,
+            height,
+        }))
+    }, [setEditorLayout])
+    const onDebugPositionUpdate = useCallback((x, y) => {
+        setDebugLayout((layout) => ({
+            ...layout,
+            x,
+            y,
+        }))
+    }, [setDebugLayout])
+    const onDebugSizeUpdate = useCallback((width, height) => {
+        setDebugLayout((layout) => ({
+            ...layout,
+            width,
+            height,
+        }))
```

Comment:
Perhaps could break out the shared functionality here into a new custom hook, something like:

```js
function useLayoutState(initialState = {}) {
  const [layout, setLayout] = useState({
    x: 0,
    y: 0,
    width: 600,
    height: 400,
    ...initialState,
  })
  const updatePosition = useCallback((x, y) => {
    setLayout((layout) => ({
      ...layout,
      x,
      y,
    }))
  }, [setLayout])
  const updateSize = useCallback((width, height) => {
    setLayout((layout) => ({
      ...layout,
      width,
      height,
    }))
  }, [setLayout])
  return useMemo(() => (
    [layout, updateSize, updatePosition]
  ), [layout, updateSize, updatePosition])
}

// …
const [debugLayout, onUpdateDebugSize, onUpdateDebugPosition] = useLayoutState()
const [editorLayout, onUpdateEditorSize, onUpdateEditorPosition] = useLayoutState()
```